### ISSUE TITLE
Xbox specific source-code is hard to find

### ISSUE THREAD

**Issue by JayFoxRox** [xqemu/xqemu#46]
Many people are scared away by the huge QEMU source tree.

However, almost all of our code is exclusively in [hw/xbox](https://github.com/xqemu/xqemu/tree/xbox/hw/xbox).

We should tell people where to look for our code, by sending them straight to that folder.
That should happen prominently in README.md

*Our web-ecosystem should also do this; and if this is not done when closing this issue, a seperate issue should be created for xqemu/xqemu.com*

**Comment by mborgerson** [comment]
Related #5

**Comment by MayeulC** [comment]
I would like to have a go at a complete re-organization of the source code. However, I will not investigate this possibility further if I do not have @mborgerson's and @JayFoxRox's (likely reluctant) approvals.

For the best result, this will require rewriting the git history (I am relatively confident in my capability of doing so while preserving most of the commit history).

## The plan
What I would like to see is the following:
* QEMU upstream as a submodule in /qemu (the commit the project was forked from, for starters)
* Today's /hw/xbox folder moved to  /xbox
* misc items.. That remains to be seen, but tidy is what I'd like it to be.

I'll sort the details out if I am to work on this, but for modifications to QEMU outside of hw/xbox (that should remain minimal anyway), we could either (or both):
* maintain patches against QEMU in a /patches folder, however that's only viable if those are temporary and to be upstreamed. If that's not the case, working with patches will be more of a burden.
* create wrappers for the QEMU code that override functions and their calls. This is a bit more involved, but I think it is pretty doable.

Of course, there are advantages and drawbacks:

### Advantages
* More clear hierarchy (xbox-specific source no longer hard to find)
* Much much more smaller code base (if you don't count upstream qemu)
* Better separation from QEMU
* Easier to sync with upstream QEMU, as we no longer need to rebase off their work, we can just update the submodule reference
* Clean git history with only qemu-pertinent stuff
* With the above item, clean Github integration of:
  * Repository stats (activity, code contributions, etc). It's not much, but can help a newcomer tell at a glance that the project is **alive and well** instead of looking at impossibly big numbers that dip towards the end
  * contributors, with easier @-mentioning
  * code history and a better view of the forks

### Drawbacks:
* It is slightly more complicated to work with a submodule, but I refuse to acknowledge it as a real problem
* It requires rewriting the history once again
  * It will break everyone's remotes and forks
  * I intend to keep the original commit dates, but the initial commit will make it seem that the project is very young if people only look at it; that will have to be explained in the readme
* It means having a more convoluted build process (nothing that's not scriptable, though). 

I can probably write a few script to help people work with the new tree (porting over their git histories, and extracting patches from the qemu tree).

If you are open enough to such a rework, I can work out a proof-of-concept.
Please note that I won't be able to work on this before the 5th of September at least (together with any other xqemu work I started since ~ 3 weeks ago; I am simply too busy).

**Comment by MayeulC** [comment]
Thanks a lot for taking the time to write this answer. I certainly understand your points.

> TL;DR: If you want to do this and prototype a potential work flow, go for it! But in general, I don't think sub-moduling QEMU core code solves the problem of this project being difficult for newcomers to navigate (in fact, I think it would make it worse). -- Barring any significant pruning or changes to the architecture of the whole thing, I think the most beneficial solution would be solid developer documentation.

At this point, I think that the only thing that could really further the discussion would be a working prototype. I think I'll start looking into this in a few weeks :wink: 
Meanwhile, feel free to merge in the QEMU upstream.
